File: frontend/js/modules/deals/toolkit.js
// Stage toolkit pane for the deal room (README §21.1, D3). Pure rendering — data comes
// from GET /api/deals/{id}/toolkit?stage=… and is computed without any AI requests.
// Anything that needs writing links to the Copilot with a prefilled question.
import { esc } from '../utils.js';

const ROLE_LABEL = { champion: 'Champion', economic_buyer: 'Economic buyer', technical_evaluator: 'Technical evaluator',
  influencer: 'Influencer', blocker: 'Blocker', user: 'User' };
const KIND_ICON = { signal: 'fa-bolt', leadership: 'fa-user-tie', hiring: 'fa-briefcase' };
const DOC_LABEL = { signal: 'Signal', news: 'News', blog: 'Blog', linkedin_post: 'LinkedIn', job_theme: 'Hiring', digest_channel: 'Digest' };

function day(d) {
  if (!d) return '';
  const dt = /^\d{4}-\d{2}-\d{2}$/.test(d) ? new Date(`${d}T00:00:00`) : new Date(d);
  return Number.isNaN(dt.getTime()) ? '' : dt.toLocaleDateString(undefined, { month: 'short', day: 'numeric', year: 'numeric' });
}

export function copilotHref(deal, q) {
  const p = new URLSearchParams({ account_id: deal.account_id, deal_id: deal.id, q });
  return `/copilot?${p}`;
}

function ask(deal, q, label = 'Draft with Copilot') {
  return `<a class="tk-ask" href="${esc(copilotHref(deal, q))}" title="${esc(q)}"><i class="fa-solid fa-wand-magic-sparkles"></i> ${esc(label)}</a>`;
}

function block(title, icon, body, extra = '') {
  return `<section class="tk-block"><h4 class="tk-h"><i class="fa-solid ${icon}"></i> ${esc(title)}${extra}</h4>${body}</section>`;
}

function empty(text) { return `<p class="tk-empty">${esc(text)}</p>`; }

function intro(d, deal) {
  const why = d.why_now.map(w => `<li class="tk-why">
      <span class="tk-why-ic ${w.kind}"><i class="fa-solid ${KIND_ICON[w.kind] || 'fa-bolt'}"></i></span>
      <div><div class="tk-why-title">${w.url ? `<a href="${esc(w.url)}" target="_blank" rel="noopener">${esc(w.title)}</a>` : esc(w.title)}</div>
        ${w.detail ? `<div class="tk-why-detail">${esc(w.detail)}</div>` : ''}
        <div class="tk-meta"><span class="tk-tag">${esc(w.label)}</span>${w.date ? ` ${esc(day(w.date))}` : ''}</div></div></li>`).join('');
  const who = d.who_to_approach.map(p => `<li class="tk-person">
      <div><a href="/profile?account=${deal.account_id}&persona_id=${p.id}" target="_blank" rel="noopener">${esc(p.name)}</a>
        <div class="tk-meta">${esc(p.title || '')}</div>
        ${p.why.length ? `<div class="tk-chips">${p.why.map(w => `<span class="tk-tag">${esc(w)}</span>`).join('')}</div>` : ''}</div>
      <div class="tk-person-ctl">
        <button type="button" class="dl-btn dl-btn-sm" data-tk-add="${p.id}" title="Add to the buying committee as an influencer — change the role later"><i class="fa-solid fa-user-plus"></i> Add</button>
        ${ask(deal, `Draft a short intro email to ${p.name}`, 'Intro email')}
      </div></li>`).join('');
  return block('Why now', 'fa-bolt', why ? `<ul class="tk-list">${why}</ul>` : empty('No recent signals, leadership moves or hiring themes for this account yet.'),
    d.why_now.length ? ' <button type="button" class="tk-link" data-tk-check="intro|trigger_noted">Mark “trigger noted”</button>' : '')
    + block('Who to approach', 'fa-bullseye', who ? `<ul class="tk-list">${who}</ul>` : empty('Everyone suggested is already on the committee.'));
}

function discovery(d, deal) {
  const banks = d.question_bank.map(b => `<details class="tk-bank" open><summary>${esc(b.name)} <span class="tk-tag">${esc(ROLE_LABEL[b.role] || b.role)}</span></summary>
      <ol>${b.questions.map(q => `<li>${esc(q)}</li>`).join('')}</ol></details>`).join('');
  const meddicc = d.meddicc.map(m => `<div class="tk-med${m.value ? ' filled' : ''}">
      <label for="tkMed_${m.key}"><strong>${esc(m.label)}</strong><span>${esc(m.hint)}</span></label>
      <textarea id="tkMed_${m.key}" data-med="${m.key}" rows="2" maxlength="1000" placeholder="${esc(m.suggested ? `Suggested: ${m.suggested}` : 'Not known yet')}">${esc(m.value)}</textarea>
      ${m.suggested && !m.value ? `<button type="button" class="tk-link" data-med-use="${m.key}" data-med-text="${esc(m.suggested)}">Use suggestion</button>` : ''}
    </div>`).join('');
  return block('MEDDICC', 'fa-clipboard-check', `<div class="tk-med-grid">${meddicc}</div>`, ` <span class="tk-score">${esc(d.meddicc_score)}</span>`)
    + block('Question bank', 'fa-circle-question', (banks || empty('Add people to the buying committee to get questions built from their KPIs, pains and objections.'))
      + `<details class="tk-bank"><summary>Questions for anyone</summary><ol>${d.generic_questions.map(q => `<li>${esc(q)}</li>`).join('')}</ol></details>`);
}

function proposal(d, deal) {
  const vm = d.value_map.map(v => `<div class="tk-value">
      <div class="tk-value-head"><strong>${esc(v.label)}</strong><span class="tk-meta">${v.evidence_count} matching item${v.evidence_count === 1 ? '' : 's'}</span></div>
      ${v.evidence.length ? `<ul>${v.evidence.map(e => `<li><span class="tk-tag">${esc(DOC_LABEL[e.doc_type] || e.doc_type)}</span>
        ${e.url ? `<a href="${esc(e.url)}" target="_blank" rel="noopener">${esc(e.title)}</a>` : esc(e.title)}</li>`).join('')}</ul>`
        : empty('No evidence at this account yet — validate the fit in discovery.')}</div>`).join('');
  const comp = d.battlecard.competitors.map(c => `<li><strong>${esc(c.name)}</strong><span class="tk-meta">named in ${c.lobs} LOB${c.lobs === 1 ? '' : 's'} · ${c.mentions} mention${c.mentions === 1 ? '' : 's'} in our data</span></li>`).join('');
  const pack = d.objection_pack.map(o => `<li class="tk-obj"><div class="tk-obj-q">“${esc(o.objection)}”<span class="tk-meta"> — ${esc(o.who)}</span></div>
      ${o.response ? `<div class="tk-obj-a"><i class="fa-solid fa-reply"></i> ${esc(o.response)}</div>` : ''}</li>`).join('');
  const bc = d.business_case;
  return block('Value map', 'fa-link', vm || empty('Pick StradIT offerings on the deal to map them to evidence.'),
    d.value_map.length ? ' <button type="button" class="tk-link" data-tk-check="proposal|offerings_mapped">Mark mapped</button>' : '')
    + block('Business case inputs', 'fa-scale-balanced', (bc.pains.length || bc.kpis.length) ? `<div class="tk-two">
        <div><strong>Pains</strong><ul>${bc.pains.map(x => `<li>${esc(x)}</li>`).join('') || '<li>—</li>'}</ul></div>
        <div><strong>KPIs they own</strong><ul>${bc.kpis.map(x => `<li>${esc(x)}</li>`).join('') || '<li>—</li>'}</ul></div></div>`
      : empty('Add committee members with call-prep to fill this in.'))
    + block('Battlecard', 'fa-chess-knight', `${comp ? `<ul class="tk-list tk-comp">${comp}</ul>` : empty('No competitors recorded for this account’s LOBs.')}
        <ul class="tk-positioning">${d.battlecard.positioning.map(p => `<li>${esc(p)}</li>`).join('')}</ul>`,
      comp ? ' <button type="button" class="tk-link" data-tk-check="proposal|competition">Mark identified</button>' : '')
    + block('Objection handling', 'fa-shield-halved', pack ? `<ul class="tk-list">${pack}</ul>` : empty('No objections recorded for the committee yet.'))
    + `<div class="tk-cta">${d.prompts.map(q => ask(deal, q, 'Draft proposal outline')).join('')}</div>`;
}

function pilot(d) {
  return block('Success criteria', 'fa-flag-checkered', `<ul class="tk-checks">${d.success_criteria.map(x => `<li>${esc(x)}</li>`).join('')}</ul>`)
    + block('Suggested 6-week plan', 'fa-calendar-days', `<ol class="tk-plan">${d.plan.map(p => `<li><span class="tk-plan-when">${esc(p.week)}<small>${esc(day(p.date))}</small></span><span>${esc(p.milestone)}</span></li>`).join('')}</ol>`)
    + block('Risk alerts', 'fa-triangle-exclamation', `<ul class="tk-risks">${d.risks.map(x => `<li>${esc(x)}</li>`).join('')}</ul>`);
}

function contract(d) {
  const chain = d.approval_chain.map((p, i) => `<li><span class="tk-step">${i + 1}</span><div><strong>${esc(p.name)}</strong>
      <span class="tk-tag">${esc(ROLE_LABEL[p.role] || p.role)}</span><div class="tk-meta">${esc(p.title || '')}${p.budget ? ` · budget: ${esc(p.budget)}` : ''}</div></div></li>`).join('');
  const lobs = d.expansion.other_lobs.map(l => `<li>${esc(l.lob_name)} <span class="tk-meta">${l.contacts} contact${l.contacts === 1 ? '' : 's'}</span></li>`).join('');
  return block('Approval chain', 'fa-stamp', chain ? `<ol class="tk-chain">${chain}</ol>` : empty('Add the buying committee to see the approval order.'))
    + block('Procurement prep', 'fa-file-signature', `<ul class="tk-checks">${d.procurement_prep.map(x => `<li>${esc(x)}</li>`).join('')}</ul>`)
    + block('Expansion map', 'fa-arrows-split-up-and-left', `<div class="tk-two">
        <div><strong>Other offerings</strong><div class="tk-chips">${d.expansion.other_offerings.map(o => `<span class="tk-tag">${esc(o.label)}</span>`).join('') || '—'}</div></div>
        <div><strong>Other lines of business</strong><ul>${lobs || '<li>—</li>'}</ul></div></div>`);
}

const RENDER = { intro, discovery, proposal, pilot, contract };

export function renderToolkit(tk, deal, stages) {
  const nav = `<div class="tk-stages" role="group" aria-label="Toolkit stage">${stages.map(s =>
    `<button type="button" class="tk-stage${s.key === (tk && tk.stage) ? ' active' : ''}${s.key === deal.stage ? ' current' : ''}" data-tk-stage="${s.key}" aria-pressed="${s.key === (tk && tk.stage)}">${esc(s.label)}</button>`).join('')}</div>`;
  if (!tk) return `${nav}<div class="tk-loading" aria-busy="true"><span></span><span></span><span></span></div>`;
  if (tk.error) return `${nav}${empty(tk.error)}`;
  return `${nav}<p class="tk-note"><i class="fa-solid fa-database"></i> Built from your account data — no AI requests used.</p>${RENDER[tk.stage](tk.data, deal)}`;
}

File: apps/sales_deals/toolkit.py
"""Stage toolkits for the deal room (apps/sales_copilot/README.md §21.1, phase D3).

Everything here is computed from data already in the DB — NO LLM requests:
  intro      why-now triggers (signals, leadership moves, hiring, news) + who to approach
  discovery  question bank per stakeholder (from their KPIs / pains / objections) + MEDDICC tracker
  proposal   value map (offering ↔ evidence), competitive battlecard, objection → response pack
  pilot      success criteria (committee KPIs), 6-week plan, risk alerts
  contract   approval chain, procurement/security prep, expansion map (other LOBs / offerings)
Anything that needs writing (an email, a proposal outline) links to the copilot with a
prefilled question, so AI requests stay explicit and quota-governed.
"""

import re
from datetime import date, datetime, timedelta, timezone
from typing import Any, Dict, List

from sqlalchemy import text

OFFERINGS = {"ai": "Applied AI", "data": "Data Analytics", "cyber": "Cybersecurity", "cloud": "Cloud & Infrastructure",
             "testing": "Automated AI Testing", "digital_assets": "Digital Assets & Blockchain"}
# Mirrors STRADIT_OFFERINGS keyword lists in frontend/js/modules/constants.js
OFFERING_KEYWORDS = {
    "ai": r"artificial intelligence|\bai\b|\bllm|generative ai|genai|machine learning|copilot|\bagents?\b|agentic",
    "data": r"data quality|data integration|predictive|forecast|dashboard|analytics|business intelligence|reporting",
    "cyber": r"cyber|security|breach|ransomware|threat|vulnerabilit|incident response|data protection|identity",
    "cloud": r"cloud|\baws\b|azure|\bgcp\b|data cent|moderni[sz]ation|multi-cloud|legacy|infrastructure",
    "testing": r"test automation|quality engineering|\bqa\b|testing|regression",
    "digital_assets": r"digital asset|tokeni[sz]|blockchain|stablecoin|crypto|custody|smart contract|on-chain",
}
MEDDICC = [("metrics", "Metrics", "What measurable outcome will they judge success by?"),
           ("economic_buyer", "Economic buyer", "Who signs off the budget?"),
           ("decision_criteria", "Decision criteria", "What will they compare vendors on?"),
           ("decision_process", "Decision process", "Steps, people and timeline to a signed decision"),
           ("identify_pain", "Identified pain", "The business pain this deal fixes"),
           ("champion", "Champion", "Who sells this internally when you're not in the room?"),
           ("competition", "Competition", "Alternatives they're considering, including doing nothing")]


def _q(s, sql: str, **kw) -> List[Dict[str, Any]]:
    return [dict(r) for r in s.execute(text(sql), kw).mappings()]


def _account(s, account_id: int) -> Dict[str, Any]:
    return _q(s, "SELECT id, key, display_name, legal_name, aliases FROM accounts WHERE id = :a", a=account_id)[0]


def _account_like(acct: Dict[str, Any]) -> List[str]:
    names = {acct["display_name"], acct["legal_name"], *(acct.get("aliases") or [])}
    return [f"%{n}%" for n in names if n and len(n) >= 3]


def _committee(s, deal_id: int) -> List[Dict[str, Any]]:
    return _q(s, """
        SELECT ds.persona_id, ds.role, ds.sentiment, coalesce(p.full_name, p.display_name) AS name, p.title,
               p.target_kpis, p.operational_pain_points, p.key_objections, p.extended_profile -> 'callprep' AS callprep,
               p.budget_authority, p.decision_authority, p.hierarchy_level, p.account_id, p.lob_id
        FROM deal_stakeholders ds JOIN personas p ON p.id = ds.persona_id WHERE ds.deal_id = :d""", d=deal_id)


def _lc(phrase: str) -> str:
    """Lower-case the first letter for mid-sentence use, keeping acronyms (AI, MLOps, KYC) intact."""
    p = (phrase or "").strip().rstrip(".")
    return p[0].lower() + p[1:] if len(p) > 1 and p[1].islower() else p


# ── Intro ─────────────────────────────────────────────────────────────────────


def intro(s, deal: Dict[str, Any]) -> Dict[str, Any]:
    acct = _account(s, deal["account_id"])
    likes = _account_like(acct)
    signals = _q(s, """SELECT title, category, details ->> 'summary' AS summary, last_seen FROM opportunity_signals
                       WHERE account_id = :a ORDER BY last_seen DESC NULLS LAST LIMIT 40""", a=deal["account_id"])
    seen, why_now = set(), []
    for x in signals:
        key = (x["summary"] or x["title"] or "")[:80]
        if key in seen:
            continue
        seen.add(key)
        why_now.append({"kind": "signal", "title": x["title"], "detail": (x["summary"] or "")[:220],
                        "date": x["last_seen"], "label": (x["category"] or "signal").replace("_", " ")})
        if len(why_now) >= 5:
            break
    moves = _q(s, """SELECT person_name, event_type, designation, effective_date, article_title, article_url, first_seen
                     FROM cxo_movements WHERE company_name ILIKE ANY(:likes) OR target_key = :k
                     ORDER BY first_seen DESC LIMIT 5""", likes=likes, k=acct["key"])
    for m in moves:
        why_now.append({"kind": "leadership", "title": f"{m['person_name']} — {m['event_type'] or 'change'} {m['designation'] or ''}".strip(),
                        "detail": m["article_title"] or "", "url": m["article_url"], "date": m["effective_date"] or m["first_seen"],
                        "label": "leadership change"})
    hiring = _q(s, """SELECT d.title, d.published_at FROM rag_documents d
                      JOIN rag_document_entities e ON e.canonical_key = d.canonical_key AND e.account_id = :a
                      WHERE d.doc_type = 'job_theme' AND d.is_current ORDER BY d.published_at DESC LIMIT 1""", a=deal["account_id"])
    if hiring:
        body = _q(s, """SELECT c.text FROM rag_documents d JOIN rag_document_chunks dc ON dc.document_id = d.id
                        JOIN rag_chunks c ON c.chunk_hash = dc.chunk_hash
                        WHERE d.title = :t AND d.is_current LIMIT 1""", t=hiring[0]["title"])
        why_now.append({"kind": "hiring", "title": hiring[0]["title"], "label": "hiring",
                        "detail": (body[0]["text"].split("\n", 1)[-1] if body else "")[:240], "date": hiring[0]["published_at"]})

    on_deal = {x["persona_id"] for x in _committee(s, deal["id"])}
    people = _q(s, """
        SELECT p.id, coalesce(p.full_name, p.display_name) AS name, p.title, p.decision_authority, p.budget_authority,
               p.tier, (p.value_proposition IS NOT NULL) AS has_callprep, p.personalized_icebreaker, p.lob_id,
               EXISTS (SELECT 1 FROM posts po WHERE po.target_key = p.key AND po.kind = 'person'
                       AND po.channel IN ('linkedin','news') AND po.first_seen > now() - interval '60 days') AS recently_active
        FROM personas p WHERE p.account_id = :a
        ORDER BY (p.decision_authority ILIKE '%final%') DESC, (p.value_proposition IS NOT NULL) DESC,
                 p.hierarchy_level NULLS LAST LIMIT 40""", a=deal["account_id"])
    offering_re = "|".join(OFFERING_KEYWORDS[o] for o in deal["offerings"] if o in OFFERING_KEYWORDS)
    ranked = []
    for p in people:
        if p["id"] in on_deal:
            continue
        score, why = 0, []
        if "final" in (p["decision_authority"] or "").lower():
            score += 3; why.append("final decision authority")
        if p["has_callprep"]:
            score += 2; why.append("call-prep ready")
        if p["recently_active"]:
            score += 2; why.append("active publicly in the last 60 days")
        if offering_re and re.search(offering_re, p["title"] or "", re.I):
            score += 2; why.append("title matches the deal's offerings")
        if deal.get("lob_id") and p["lob_id"] == deal["lob_id"]:
            score += 2; why.append("in the deal's line of business")
        ranked.append({**p, "score": score, "why": why})
    ranked.sort(key=lambda x: -x["score"])
    return {"why_now": why_now[:8], "who_to_approach": ranked[:6],
            "prompts": [f"Draft a short intro email to {p['name']}" for p in ranked[:2]]}


# ── Discovery ─────────────────────────────────────────────────────────────────


def discovery(s, deal: Dict[str, Any]) -> Dict[str, Any]:
    committee = _committee(s, deal["id"])
    banks = []
    for p in committee:
        qs = []
        for pain in (p["operational_pain_points"] or [])[:3]:
            qs.append(f"How is {_lc(pain)} affecting your team today, and what have you already tried?")
        for kpi in (p["target_kpis"] or [])[:2]:
            qs.append(f"How do you measure {_lc(kpi)} today, and where does it need to be in 12 months?")
        objections = ((p["callprep"] or {}).get("objections") or [{"objection": o} for o in (p["key_objections"] or [])])[:2]
        for o in objections:
            if isinstance(o, dict) and o.get("objection"):
                qs.append(f"Pre-empt: \"{o['objection']}\" — what would you need to see to be comfortable on that?")
        if p["role"] == "economic_buyer":
            qs.append("What has to be true for this to get budget this year, and who else weighs in?")
        if p["role"] == "technical_evaluator":
            qs.append("What would a technical evaluation need to prove, and on what timeline?")
        if not qs:
            qs = ["What are your top priorities for the next two quarters?",
                  "Where do current processes slow your team down the most?",
                  "How are decisions like this usually made and approved?"]
        banks.append({"persona_id": p["persona_id"], "name": p["name"], "role": p["role"], "questions": qs[:6]})
    q = deal.get("qualification") or {}
    roles = {x["role"]: x["name"] for x in committee}
    pains = sorted({pp for x in committee for pp in (x["operational_pain_points"] or [])})
    competitors = _competitors(s, deal)
    meddicc = []
    for key, label, hint in MEDDICC:
        suggested = {"economic_buyer": roles.get("economic_buyer"), "champion": roles.get("champion"),
                     "identify_pain": "; ".join(pains[:3]) or None,
                     "competition": ", ".join(c["name"] for c in competitors[:3]) or None}.get(key)
        meddicc.append({"key": key, "label": label, "hint": hint, "value": q.get(key) or "", "suggested": suggested or ""})
    filled = sum(1 for m in meddicc if m["value"] or m["suggested"])
    return {"question_bank": banks, "meddicc": meddicc, "meddicc_score": f"{filled}/{len(meddicc)}",
            "generic_questions": ["What happens if you do nothing this year?", "Who else is affected by this problem?",
                                  "What does the approval process look like for a project this size?"]}


# ── Proposal ──────────────────────────────────────────────────────────────────


def _competitors(s, deal: Dict[str, Any]) -> List[Dict[str, Any]]:
    rows = _q(s, "SELECT competitors FROM lobs WHERE account_id = :a AND competitors IS NOT NULL", a=deal["account_id"])
    counts: Dict[str, int] = {}
    for r in rows:
        for c in (r["competitors"] or []):
            counts[str(c)] = counts.get(str(c), 0) + 1
    return [{"name": k, "lobs": v} for k, v in sorted(counts.items(), key=lambda kv: -kv[1])[:6]]


def proposal(s, deal: Dict[str, Any]) -> Dict[str, Any]:
    value_map = []
    for o in deal["offerings"] or []:
        if o not in OFFERING_KEYWORDS:
            continue
        ev = _q(s, """
            SELECT DISTINCT ON (d.canonical_key) d.title, d.doc_type, d.url, d.published_at
            FROM rag_documents d JOIN rag_document_entities e ON e.canonical_key = d.canonical_key AND e.account_id = :a
            JOIN rag_document_chunks dc ON dc.document_id = d.id JOIN rag_chunks c ON c.chunk_hash = dc.chunk_hash
            WHERE d.is_current AND d.deleted_at IS NULL AND d.doc_type IN ('signal','news','blog','linkedin_post','job_theme','digest_channel')
              AND c.text ~* :re
            ORDER BY d.canonical_key, d.published_at DESC NULLS LAST LIMIT 40""", a=deal["account_id"], re=OFFERING_KEYWORDS[o])
        ev.sort(key=lambda x: x["published_at"] or datetime.min.replace(tzinfo=timezone.utc), reverse=True)
        value_map.append({"offering": o, "label": OFFERINGS[o], "evidence": ev[:4], "evidence_count": len(ev)})
    committee = _committee(s, deal["id"])
    pack = []
    for p in committee:
        cp = p["callprep"] or {}
        for ob in (cp.get("objections") or [{"objection": o, "counter": ""} for o in (p["key_objections"] or [])])[:3]:
            if isinstance(ob, dict) and ob.get("objection"):
                pack.append({"who": p["name"], "role": p["role"], "objection": ob["objection"], "response": ob.get("counter") or ""})
    acct = _account(s, deal["account_id"])
    comp = _competitors(s, deal)
    for c in comp:
        c["mentions"] = s.execute(text("""
            SELECT count(DISTINCT d.id) FROM rag_documents d
            JOIN rag_document_entities e ON e.canonical_key = d.canonical_key AND e.account_id = :a
            JOIN rag_document_chunks dc ON dc.document_id = d.id JOIN rag_chunks ch ON ch.chunk_hash = dc.chunk_hash
            WHERE d.is_current AND ch.text ILIKE :n"""), {"a": deal["account_id"], "n": f"%{c['name'].split(' ')[0]}%"}).scalar()
    kpis = sorted({k for p in committee for k in (p["target_kpis"] or [])})
    pains = sorted({k for p in committee for k in (p["operational_pain_points"] or [])})
    return {"value_map": value_map, "battlecard": {"competitors": comp,
            "positioning": [f"Anchor on {acct['display_name']}'s own priorities (see value map) rather than feature comparisons.",
                            "Offer a scoped pilot with agreed success criteria to lower switching risk.",
                            "Bring a reference architecture that fits their existing stack (see LOB technologies)."]},
            "objection_pack": pack, "business_case": {"pains": pains[:5], "kpis": kpis[:5]},
            "prompts": [f"Draft a one-page proposal outline for {acct['display_name']} covering "
                        + ", ".join(OFFERINGS[o] for o in deal["offerings"] if o in OFFERINGS)]}


# ── Pilot ─────────────────────────────────────────────────────────────────────


def pilot(s, deal: Dict[str, Any]) -> Dict[str, Any]:
    committee = _committee(s, deal["id"])
    kpis = sorted({k for p in committee for k in (p["target_kpis"] or [])})
    criteria = [f"Agreed baseline and target for: {k}" for k in kpis[:4]] or \
        ["Agree 2-3 measurable outcomes with the champion before kickoff"]
    start = date.today() + timedelta(days=7)
    plan = [{"week": "Week 0", "date": start.isoformat(), "milestone": "Kickoff: success criteria, data access, owners signed off"},
            {"week": "Week 1-2", "date": (start + timedelta(days=14)).isoformat(), "milestone": "Build + first results on a narrow use case"},
            {"week": "Week 3-4", "date": (start + timedelta(days=28)).isoformat(), "milestone": "Expand to target users; weekly check-ins"},
            {"week": "Week 5", "date": (start + timedelta(days=35)).isoformat(), "milestone": "Measure against criteria; security review"},
            {"week": "Week 6", "date": (start + timedelta(days=42)).isoformat(), "milestone": "Results readout with economic buyer; decision"}]
    risks = []
    last = s.execute(text("SELECT max(created_at) FROM deal_activity WHERE deal_id = :d"), {"d": deal["id"]}).scalar()
    if last and (datetime.now(timezone.utc) - last).days > 14:
        risks.append(f"No activity for {(datetime.now(timezone.utc) - last).days} days")
    for p in committee:
        if p["sentiment"] == "negative" or p["role"] == "blocker":
            risks.append(f"{p['name']} is a blocker / negative — plan how to address their concerns")
        moved = s.execute(text("""SELECT 1 FROM cxo_movements WHERE lower(person_name) = lower(:n)
                                  AND first_seen > now() - interval '90 days' LIMIT 1"""), {"n": p["name"]}).scalar()
        if moved:
            risks.append(f"{p['name']} ({p['role'].replace('_', ' ')}) had a recent leadership change — re-confirm sponsorship")
    if "champion" not in {p["role"] for p in committee}:
        risks.append("No champion — pilots without an internal champion rarely convert")
    return {"success_criteria": criteria, "plan": plan, "risks": risks or ["No risks detected from the data"]}


# ── Contract ──────────────────────────────────────────────────────────────────


def contract(s, deal: Dict[str, Any]) -> Dict[str, Any]:
    committee = _committee(s, deal["id"])
    order = {"economic_buyer": 0, "champion": 1, "technical_evaluator": 2, "influencer": 3, "user": 4, "blocker": 5}
    chain = sorted(committee, key=lambda p: (0 if "sign-off" in (p["budget_authority"] or "").lower() else 1,
                                            order.get(p["role"], 9), p["hierarchy_level"] or 9))
    lobs = _q(s, """SELECT l.id, l.lob_name, count(p.id) AS contacts FROM lobs l LEFT JOIN personas p ON p.lob_id = l.id
                    WHERE l.account_id = :a AND l.id IS DISTINCT FROM :lob GROUP BY l.id ORDER BY 3 DESC, 2 LIMIT 5""",
             a=deal["account_id"], lob=deal.get("lob_id"))
    return {"approval_chain": [{"name": p["name"], "title": p["title"], "role": p["role"], "budget": p["budget_authority"]} for p in chain],
            "procurement_prep": ["Security questionnaire and InfoSec review pack ready", "Data processing / residency terms",
                                 "Vendor onboarding documents (insurance, certifications)", "Commercial terms and approval thresholds",
                                 "Legal redlines owner and turnaround agreed"],
            "expansion": {"other_offerings": [{"key": k, "label": v} for k, v in OFFERINGS.items() if k not in (deal["offerings"] or [])],
                          "other_lobs": lobs}}


BUILDERS = {"intro": intro, "discovery": discovery, "proposal": proposal, "pilot": pilot, "contract": contract}
